clouddriver-dcos/src/main/groovy/com/netflix/spinnaker/clouddriver/dcos/deploy/util/DeployDcosServerGroupDescriptionToAppMapper.groovy

```diff
@@ -124,28 +126,37 @@ class DeployDcosServerGroupDescriptionToAppMapper {
         parsedConstraints
     }
 
-    public Map<String, String> parsePortMappingLabels(String appId, List<DeployDcosServerGroupDescription.ServiceEndpoint> serviceEndpoints) {
-        Map<String, String> parsedMap = [:]
+    private Tuple parsePortMappingLabels(String appId, DeployDcosServerGroupDescription.ServiceEndpoint serviceEndpoint, int counter) {
+        Map<String, String> parsedLabels = serviceEndpoint.labels.clone()
 
-        int counter = 0
+        List<String> vipKeys = parsedLabels.keySet().stream().filter({ key -> key.matches(VIP_PATTERN) })
+                .collect(Collectors.toList())
 
-        serviceEndpoints.stream().filter({
-            serviceEndpoint -> serviceEndpoint.isLoadBalanced
-        }).forEach({
-            serviceEndpoint -> parsedMap.put("VIP_${counter++}", "${appId}:${serviceEndpoint.port}")
+        vipKeys.stream().forEach({
+            key ->
+                String vip = parsedLabels.remove(key)
+                parsedLabels.put("VIP_${counter++}".toString(), vip)
         })
 
-        parsedMap
+        if (serviceEndpoint.loadBalanced && vipKeys.isEmpty()) {
+            parsedLabels.put("VIP_${counter++}".toString(), "${appId}:${serviceEndpoint.port}".toString())
+        }
+
+        return [counter, parsedLabels]
     }
 
     public List<Port> parsePortMappings(String appId, List<DeployDcosServerGroupDescription.ServiceEndpoint> serviceEndpoints) {
+
+        int counter = 0
```

Comment:
I believe the counter really just needs to be the index of the port mapping item on which the label is being added. Here's an example DC/OS json when the first port is load balanced, the second is not, and the third is:

      "portMappings": [
        {
          "containerPort": 40000,
          "protocol": "tcp",
          "name": "1",
          "servicePort": 10106,
          "labels": {
            "VIP_0": "/alva/bar/foo-vol-test-v000:40000"
          }
        },
        {
          "containerPort": 45000,
          "protocol": "tcp",
          "name": "2",
          "servicePort": 10116
        },
        {
          "containerPort": 50000,
          "protocol": "tcp",
          "name": "3",
          "servicePort": 10117,
          "labels": {
            "VIP_2": "/alva/bar/foo-vol-test-v000:50000"
          }
        }
      ],

You could use java 8 to map with an index (i.e. start the stream with `IntStream.range(0, serviceEndpoints.size()).map(index -> serviceEndpoints.get(index))....`) 

or use groovy withIndex() - `serviceEndpoints.withIndex().collect { element, index ->...`